docs/style-guide.md

```diff
@@ -0,0 +1,1348 @@
+
+C codes should be C99 compliant and compiled with the `-std=c99` flag,
+and C++ codes should be C++11 compliant and compiled with the `-std=c++11` flag
+(or later).
+
+Avoid features present only in C but not in C++. That is, C code should compile with either C or C++ compiler.
+
+> Microsoft's C compiler doesn't support C99, so code must use C++11.
+
+Header Files
+============
+
+Self-contained Headers
+----------------------
+
+Header files should be self-contained and end in `.h` for C and in `.hh` for C++.
+Files that are meant for textual inclusion, but are not headers, should end in `.inc`.
+
+> Google uses `.cc` for C++ source files, but `.h` for C++ header files.
+> However, it is useful to have a distinction between C and C++ headers.
+> Therefore, we use `.c` and `.h` for C and `.cc` and `.hh` for C++.
+>
+> Trilinos uses another common convention of `.cpp` and `.hpp`.
+> However, in a long list of files, this puts a lot of p's on the screen.
+> The `.cc` and `.hh` endings are shorter and cleaner.
+
+\#define Guards
+--------------
+
+All header files should have `#define` guards to prevent multiple inclusion.
+The format of the symbol name should be `<PROJECT>_<FILE>_H` for C
+and `<PROJECT>_<FILE>_HH` for C++.
+
+```
+#!cpp
+
+#ifndef PLASMA_BLAS_H
+#define PLASMA_BLAS_H
+
+...
+
+#endif // PLASMA_BLAS_H
+```
+
+> Google uses an underscore at the end.
+> Trilinos does not use underscores.
+> In the case of header guards, beginning/ending underscores seem pointless.
+> Underscores in front of a name are reserved for system-level headers.
+
+extern C
+--------
+
+C library headers should add extern "C" around function definitions, to allow C++ codes to use the library.
+
+```
+#!cpp
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+...
+
+#ifdef __cplusplus
+}  // extern "C"
+#endif
+```
+
+Forward Declarations
+--------------------
+
+Avoid using forward declarations, i.e., declarations of classes, functions,
+or templates without associated definitions.
+Instead, `#include` the headers you need. This limits the number of places that
+have to be modified if the members change.
+Check [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Forward_Declarations)
+to see why.
+
+Inline Functions
+----------------
+
+Mark small functions as inline, specifically those that serve as macro replacements.
+Although the compiler will identify on its own functions
+suitable for inlining, it is a good idea hint it directly.
+Define as inline member accessors and functions that would otherwise be macros,
+e.g., address arithmetic functions and alike.
+Check [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Inline_Functions)
+for further guidelines.
+
+Names and Order of Included Header Files
+----------------------------------------
+
+Use the following order of inclusion:
+
+* your project headers,
+* standard headers,
+* other libraries' headers.
+
+For example, the include section might look like this:
+
```

Comment:
Project headers can use both quotes and angle brackets, right?

Reply:
Google recommends only standard library and system headers use angle brackets.
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
The PLASMA style guide seems silent on it, except for this example. Most PLASMA includes use quotes, but some use angle brackets.
```
plasma> git grep '#include "plasma' | wc
   1071    2162   49464
plasma> git grep '#include <plasma' | wc
    138     276    7005
```
Similarly in SLATE, > 90% use quotes.